clean_dockerhub.py

```diff
@@ -0,0 +1,277 @@
+                data = await resp.json()
+
+            results = data.get("results", [])
+            if not results:
+                break
+
+            for t in results:
+                # Filter tags older than threshold
+                updated = parse_iso_datetime(t["last_updated"])
+                if updated < threshold:
+                    tags.append(Tag(name=t["name"], date=updated, deleted=False))
+
+            if not data.get("next"):
+                break
+
+            page += 1
+            status.update(status_msg.format(n_pages=page))
+
+            if 0 < max_tags <= len(tags):
+                break
+    console.log(f"[bold green]Total tags fetched: {len(tags)}")
+    return tags
+
+
+async def delete_tag(
+    session: aiohttp.ClientSession,
+    namespace: str,
+    repo: str,
+    tag: Tag,
+    token: str,
+    semaphore: asyncio.Semaphore,
+) -> int:
+    """
+    Delete a single tag from the repository.
+    If a 429 code is returned by Docker Hub, a TooManyRequestsError is raised.
+    Otherwise, the status code is returned.
+    """
+    url = f"{DOCKERHUB_API}/repositories/{namespace}/{repo}/tags/{tag.name}/"
+    headers = {"Authorization": f"JWT {token}"}
+
+    async with semaphore:
+        async with session.delete(url, headers=headers) as resp:
+            if resp.status == 204:
+                tag.deleted = True
+            elif resp.status == 429:
+                # Rate limit hit
+                raise TooManyRequestsError(f"{namespace}/{repo}:{tag}")
+            return resp.status
+
+
+async def get_old_tags(
+    namespace: str,
+    repo: str,
+    token: str,
+    min_age_in_days: int,
+    max_number_of_pages: int,
+) -> list[Tag]:
+    """
+    Calculates the time threshold, opens a aiohttp.ClientSession session and then calls `fetch_old_tags`.
+    """
+    threshold = datetime.now() - timedelta(days=min_age_in_days)
+
+    conn = aiohttp.TCPConnector(limit_per_host=10)
+    async with aiohttp.ClientSession(connector=conn) as session:
+        # 2) Fetch all tags
+        old_tags = await fetch_old_tags(
+            session, namespace, repo, threshold, token, max_number_of_pages
+        )
+        return old_tags
+
+
+async def delete_tags(
+    namespace: str,
+    repo: str,
+    token: str,
+    tags_to_delete: list[Tag],
+) -> bool:
+    """
+    Delete given tags from a repository, always with 3 Http requests in parallel.
+    The method creates an aiohttp.ClientSession session, then creates all async delete tasks.
+    Whenever one tasks raises the TooManyRequestsError exception, all other tasks are cancelled,
+    and the method returns True. If no such exception is raised, the method returns False.
+    """
+
+    # Higher values will run deletion faster, but have higher risk of getting a 429,
+    # see https://docs.docker.com/docker-hub/usage/#abuse-rate-limit
+    # Tests have shown that max deletion rate without running into rate limits is ~650 Tags/min.
```

Comment:
How does the implementation `asyncio.Semaphore(2)` relate to the comment `~650 Tags/min.`?